### PR TITLE
[api] Memory optimizations for API frame helper buffering

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

This PR optimizes the API frame helper's buffering logic to reduce flash usage and eliminate code duplication.

**Changes:**
1. **Pass total write length to `write_raw_`**: Callers now provide the total write length they already know, eliminating redundant iteration through iovecs to calculate it.
2. **Eliminate code duplication in partial send handling**: Refactored the partial send buffering logic to reuse the existing `buffer_data_from_iov_` function with an offset parameter.
3. **Optimize SendBuffer memory allocation**: Replaced `std::vector<uint8_t>` with `std::unique_ptr<uint8_t[]>` since buffers are never resized after creation. This reduces the SendBuffer struct size from 16 to 8 bytes on 32-bit systems.

**Results:**
- Total flash savings: 340 bytes
  - 4 bytes saved from eliminating redundant length calculations
  - 172 bytes saved from eliminating code duplication
  - 164 bytes saved from SendBuffer optimization

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):** N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** N/A (internal implementation change)

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# No configuration changes required - internal optimization only
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

**Additional testing:**
- Verified the refactored `buffer_data_from_iov_` function with comprehensive unit tests covering:
  - Various offset positions (within segments, at boundaries, past segments)
  - Edge cases (empty buffers, single segments, tiny segments)
  - All tests pass confirming correct behavior

**Memory impact:**
- RAM usage: No change (15492 bytes)
- Flash usage: Reduced from 523544 to 523204 bytes (-340 bytes)

### CODE REVIEW

**Comment by bdraco** [comment]
test
```cpp
#include <vector>
#include <deque>
#include <iostream>
#include <sys/uio.h>
#include <cstring>
#include <cassert>
#include <memory>

// Simulate the SendBuffer structure with unique_ptr
struct SendBuffer {
    std::unique_ptr<uint8_t[]> data;
    uint16_t size{0};
    uint16_t offset{0};
};

// Test version of the function using unique_ptr
void buffer_data_from_iov(const struct iovec *iov, int iovcnt, uint16_t total_write_len, 
                          uint16_t offset, SendBuffer& buffer) {
    buffer.size = total_write_len - offset;
    buffer.data = std::make_unique<uint8_t[]>(buffer.size);
    
    uint16_t to_skip = offset;
    uint16_t write_pos = 0;
    
    for (int i = 0; i < iovcnt; i++) {
        if (to_skip >= iov[i].iov_len) {
            // Skip this entire segment
            to_skip -= static_cast<uint16_t>(iov[i].iov_len);
        } else {
            // Include this segment (partially or fully)
            const uint8_t *src = reinterpret_cast<uint8_t *>(iov[i].iov_base) + to_skip;
            uint16_t len = static_cast<uint16_t>(iov[i].iov_len) - to_skip;
            std::memcpy(buffer.data.get() + write_pos, src, len);
            write_pos += len;
            to_skip = 0;
        }
    }
}

void print_buffer(const uint8_t* data, uint16_t size, const char* name) {
    std::cout << name << ": ";
    for (uint16_t i = 0; i < size; i++) {
        std::cout << (int)data[i] << " ";
    }
    std::cout << std::endl;
}

void test_case(const char* name, const struct iovec* iov, int iovcnt, 
               uint16_t total_len, uint16_t offset, 
               const std::vector<uint8_t>& expected) {
    std::cout << "\nTest: " << name << std::endl;
    std::cout << "Total length: " << total_len << ", Offset: " << offset << std::endl;
    
    SendBuffer buffer;
    buffer_data_from_iov(iov, iovcnt, total_len, offset, buffer);
    
    print_buffer(expected.data(), expected.size(), "Expected");
    print_buffer(buffer.data.get(), buffer.size, "Got");
    
    bool pass = (buffer.size == expected.size() && 
                 std::memcmp(buffer.data.get(), expected.data(), buffer.size) == 0);
    
    if (pass) {
        std::cout << "PASS" << std::endl;
    } else {
        std::cout << "FAIL" << std::endl;
        assert(false);
    }
}

int main() {
    // Test data
    uint8_t data1[] = {1, 2, 3, 4, 5};
    uint8_t data2[] = {6, 7, 8};
    uint8_t data3[] = {9, 10, 11, 12};
    
    struct iovec iov[3] = {
        {data1, sizeof(data1)},  // 5 bytes
        {data2, sizeof(data2)},  // 3 bytes
        {data3, sizeof(data3)}   // 4 bytes
    };
    // Total: 12 bytes
    
    // Test 1: No offset (buffer everything)
    test_case("No offset", iov, 3, 12, 0, 
              {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12});
    
    // Test 2: Offset within first segment
    test_case("Offset 2 (within first segment)", iov, 3, 12, 2,
              {3, 4, 5, 6, 7, 8, 9, 10, 11, 12});
    
    // Test 3: Offset at segment boundary
    test_case("Offset 5 (at first segment boundary)", iov, 3, 12, 5,
              {6, 7, 8, 9, 10, 11, 12});
    
    // Test 4: Offset in middle segment
    test_case("Offset 7 (in second segment)", iov, 3, 12, 7,
              {8, 9, 10, 11, 12});
    
    // Test 5: Offset at last segment
    test_case("Offset 8 (at second segment boundary)", iov, 3, 12, 8,
              {9, 10, 11, 12});
    
    // Test 6: Large offset
    test_case("Offset 10 (in last segment)", iov, 3, 12, 10,
              {11, 12});
    
    // Test 7: Offset equals total length (nothing to buffer)
    test_case("Offset equals total length", iov, 3, 12, 12,
              {});
    
    // Test 8: Single iovec with offset
    struct iovec single_iov[] = {{data1, sizeof(data1)}};
    test_case("Single iovec with offset", single_iov, 1, 5, 3,
              {4, 5});
    
    // Test 9: Empty iovec array
    test_case("Empty iovec array", nullptr, 0, 0, 0,
              {});
    
    // Test edge case: very small segments
    uint8_t tiny1[] = {1};
    uint8_t tiny2[] = {2};
    uint8_t tiny3[] = {3};
    struct iovec tiny_iov[3] = {
        {tiny1, 1},
        {tiny2, 1},
        {tiny3, 1}
    };
    
    test_case("Tiny segments, offset 0", tiny_iov, 3, 3, 0,
              {1, 2, 3});
    
    test_case("Tiny segments, offset 1", tiny_iov, 3, 3, 1,
              {2, 3});
    
    test_case("Tiny segments, offset 2", tiny_iov, 3, 3, 2,
              {3});
    
    std::cout << "\nAll tests passed!" << std::endl;
    
    return 0;
}
```

File: esphome/components/api/api_frame_helper.cpp [hunk 1]
```diff
@@ -77,32 +77,45 @@ APIError APIFrameHelper::loop() {
...
   for (int i = 0; i < iovcnt; i++) {
-    const uint8_t *data = reinterpret_cast<uint8_t *>(iov[i].iov_base);
-    buffer.data.insert(buffer.data.end(), data, data + iov[i].iov_len);
+    if (to_skip >= iov[i].iov_len) {
+      // Skip this entire segment
+      to_skip -= static_cast<uint16_t>(iov[i].iov_len);
+    } else {
+      // Include this segment (partially or fully)
+      const uint8_t *src = reinterpret_cast<uint8_t *>(iov[i].iov_base) + to_skip;
+      uint16_t len = static_cast<uint16_t>(iov[i].iov_len) - to_skip;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/api/api_frame_helper.cpp:96]: The math is guaranteed to work out correctly if the caller provides accurate values.  If they didn't we are broken either way.  We already set the iov lengths manually, this is no different.

File: esphome/components/api/api_frame_helper.cpp [hunk 2]
```diff
@@ -678,12 +673,13 @@ APIError APINoiseFrameHelper::write_protobuf_packets(ProtoWriteBuffer buffer, st
     buf_start[2] = static_cast<uint8_t>(mbuf.size);
 
     // Add iovec for this encrypted packet
-    this->reusable_iovs_.push_back(
-        {buf_start, static_cast<size_t>(3 + mbuf.size)});  // indicator + size + encrypted data
+    size_t packet_len = static_cast<size_t>(3 + mbuf.size);  // indicator + size + encrypted data
+    this->reusable_iovs_.push_back({buf_start, packet_len});
+    total_write_len += packet_len;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/api/api_frame_helper.cpp:678]: defensive checks here is just wasting flash space for no real benefit since the packet is already limited to UINT16_MAX

File: esphome/components/api/api_frame_helper.cpp [hunk 3]
```diff
@@ -1064,12 +1061,13 @@ APIError APIPlaintextFrameHelper::write_protobuf_packets(ProtoWriteBuffer buffer
         .encode_to_buffer_unchecked(buf_start + header_offset + 1 + size_varint_len, type_varint_len);
 
     // Add iovec for this packet (header + payload)
-    this->reusable_iovs_.push_back(
-        {buf_start + header_offset, static_cast<size_t>(total_header_len + packet.payload_size)});
+    size_t packet_len = static_cast<size_t>(total_header_len + packet.payload_size);
+    this->reusable_iovs_.push_back({buf_start + header_offset, packet_len});
+    total_write_len += packet_len;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/api/api_frame_helper.cpp:1066]: defensive checks here is just wasting flash space for no real benefit since the packet is already limited to UINT16_MAX

**Review by jesserockz — APPROVED** [review]

**Comment by bdraco** [comment]
Thanks
